### PR TITLE
Expand LXD driver functionality

### PR DESCRIPTION
**Pull request by wilmardo**
As the title states I expanded the LXD driver. The default Docker driver doesn't cut it for me because of the missing systemd. Took a try with [gdraheim/docker-systemctl-replacement](https://github.com/gdraheim/docker-systemctl-replacement) but it comes short sometimes, not suitable for stable testing. 

## What has changed:
* Mainly more mappings of the lxd_container module to allow more flexible usage
* The lxd_container option `profiles` defaults to the default profile when not provided([source](https://lxd.readthedocs.io/en/latest/profiles/)) therefore now standard omitted
* The prepare playbook needed work since Python is not always installed in LXC containers. Also installation of some basic packages (based on the Dockerfile in the repo). Which for Gentoo took some bash magic for the Ansible required gentoolkit.
* Moved the create, destroy and prepare playbooks to the `provisioner/ansible/playbooks` directory
* Fixed a little unrelated typo in molecule init template docs

## Testing done:
* Ran `tox -e py27-ansible26-functional -- -k lxd -v` which passed successfully
* Created a test repo and did a travis build against all the aliases on the [official LXC repo](https://uk.images.linuxcontainers.org/) see [the travis build](https://travis-ci.org/wilmardo/molecule-pr-lxd-rework/builds/415664480). Only UbuntuCore and Arch failed. UbuntuCore is special so that is fine, Arch fails because of [this issue](https://github.com/lxc/lxd/issues/4071).

## Two points for discussion:
* Do I need to update the LXD driver tests to reflect these new changes? I thought about it but in my opinion it doesn't add much since it only implements the lxd_container module
* The lxd_container option `source` is a dict in the lxd_container module but not in molecule. I thought about changing this but it would break backwards compatibility since alias moves into this dict

### CODE REVIEW

File: molecule/cookiecutter/scenario/driver/lxd/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/prepare.yml [hunk 1]
```diff
@@ -1,11 +0,0 @@
----
```
Comment (retr0h) [inline molecule/cookiecutter/scenario/driver/lxd/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/prepare.yml]: Why was this removed?

File: molecule/driver/lxd.py [hunk 2]
```diff
@@ -38,6 +38,17 @@ class LXD(base.Base):
...
+          - name: instance
+            url: https://127.0.0.1:8443
+            cert_file: "/root/.config/lxc/client.crt"
+            key_file: "/root/.config/lxc/client.key"
+            trust_password: password
+            mode: pull|local
+            server: https://images.linuxcontainers.org
+            alias: ubuntu/xenial/amd64
+            profiles:
+              - default
```
Comment (retr0h) [inline molecule/driver/lxd.py]: Since you have implemented this API for molecule.yml we need to validate the schema.  Can you update the v2 schema similar to what we did with dockder and vagrant.

https://github.com/metacloud/molecule/blob/master/molecule/model/schema_v2.py#L596
https://github.com/metacloud/molecule/blob/master/molecule/model/schema_v2.py#L862

And we will need to update tests with working and failing tests:

https://github.com/metacloud/molecule/blob/master/test/unit/model/v2/test_platforms_section.py

File: molecule/cookiecutter/scenario/driver/lxd/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/prepare.yml (hunk 1 above)
Comment (wilmardo) [reply molecule/cookiecutter/scenario/driver/lxd/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/prepare.yml]: It is replaced by the expanded `molecule/provisioner/ansible/playbooks/lxd/prepare.yml` as seen a bit further below.

Comment (retr0h) [reply molecule/cookiecutter/scenario/driver/lxd/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/prepare.yml]: Oh, I see I had missed that.  The prepare you created is similar to our Dockerfile setup where we ensure the docker container has the necessary dependencies. 👍

**Comment by wilmardo** [comment]
> Do I need to update the LXD driver tests to reflect these new changes? I thought about it but in my opinion it doesn't add much since it only implements the lxd_container module

Looked into this with help of your comment, I think I implemented it all. Most if it was similar to the vagrant|docker implementation, only the devices argument which has a nested dict where the keys can differ was difficult. See the [lxd API docs](https://github.com/lxc/lxd/blob/master/doc/rest-api.md#post-1) and the [lxd devices docs](https://lxd.readthedocs.io/en/latest/containers/#devices-configuration)

> The lxd_container option source is a dict in the lxd_container module but not in molecule. I thought about changing this but it would break backwards compatibility since alias moves into this dict

Had some inspiration and moved the source options to a dict with some help of Jinja2. This is the same as the lxd_container module and keeps things more easy for the end user. The part about breaking backward compatibility was not true since there was no variables before (used a custom create.yml myself, that got me confused).

File: molecule/model/schema_v2.py [hunk 3]
```diff
@@ -710,6 +710,85 @@ def pre_validate_base_schema(env, keep_string):
...
+                                'pull',
+                                'local',
+                            ],
+                        },
+                        'server': {
+                            'type': 'string',
+                        },
+                        'protocol': {
+                            'type': 'string',
+                            'allowed': [
```
Comment (wilmardo) [inline molecule/model/schema_v2.py]: I made the schema as specific as possible, these are the only values valid for the LXD api. There are now allowed values ([source](https://github.com/lxc/lxd/blob/master/doc/rest-api.md#post-1)).

File: molecule/model/schema_v2.py [hunk 4]
```diff
@@ -710,6 +710,85 @@ def pre_validate_base_schema(env, keep_string):
...
+                            ],
+                        },
+                        'alias': {
+                            'type': 'string',
+                        },
+                    },
+                },
+                'architecture': {
+                    'type': 'string',
+                    'allowed': [
```
Comment (wilmardo) [inline molecule/model/schema_v2.py]: Same here ([source](https://github.com/lxc/lxd/blob/master/doc/rest-api.md#post-1))

File: molecule/model/schema_v2.py [hunk 5]
```diff
@@ -710,6 +710,85 @@ def pre_validate_base_schema(env, keep_string):
...
+                        'x86_64',
+                        'i686',
+                    ],
+                },
+                'devices': {
```
Comment (wilmardo) [inline molecule/model/schema_v2.py]: Please take a good look at the devices schema, that is the one I struggled with since I am not familiar with the markup. Might be not 100% correct

**Comment by wilmardo** [comment]
Is there a reason why there is a `test/resources/playbooks/lxd`? Couldn't we just point to the `molecule/provisioner/ansible/playbooks/lxd` in molecule.yml for testing? This would dedup these playbooks.

**Comment by wilmardo** [comment]
@retr0h Do you know when you have time to review the latest changes? In no means I want to rush you, just curious if everything is alright now :)

**Comment by retr0h** [comment]
@wilmardo the model unit tests are failing.

**Comment by retr0h** [comment]
Also, run `tox -e format` to reformat the code with yapf.

**Comment by retr0h** [comment]
@wilmardo v2 model unit tests still failing

```
test/unit/model/v2/test_platforms_section.py::test_platforms_lxd[_model_platforms_lxd_section_...] FAILED                                       [ 62%]
test/unit/model/v2/test_platforms_section.py::test_platforms_lxd_has_errors[_model_platforms_docker_error...] FAILED                            [ 64%]
```

**Comment by wilmardo** [comment]
@retr0h Yeah sadly I know, I struggled with it yesterday but there seems to be an issue on my own laptop since the master branch is failing too. Hopefully this evening I have time to retry on my desktop on which I know it worked before.
Any tips on debugging the tox errors? They seem very non descriptive to me (first time using tox).
Also are there any useful tox commands I (or any other contributor) need to know? I might expand the Testing section of the docs some more, for example only yesterday I found out the schema is Cerberus. Also the LXD test commands are not longer valid (v1 legacy?).

**Comment by retr0h** [comment]
I suggest just running py.test.

    $ py.test test/unit/model/v2/ -vvv

You can then see which test failed.

**Comment by wilmardo** [comment]
Fixed the unit tests! Pytest helped a lot, I will try to create a PR for some more documentation about testing. 
The most of the issue was that in `test_platforms_lxd_has_errors` the items in the source variable are alphabetically sorted and source itself becomes a dict in a list for some reason.

Thanks for the pointers!

**Comment by retr0h** [comment]
Merging.  @wilmardo is our resident LXD driver community  member.
